runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -1930,13 +1930,624 @@ TR::Register *J9::Power::TreeEvaluator::anewArrayEvaluator(TR::Node *node, TR::C
+                                         CONSTANT64(0x0000001FFFFFFFE0));
+         break;
+      case 16:
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldic, node, targetReg, sourceReg, 4,
+                                         CONSTANT64(0x0000000FFFFFFFF0));
+         break;
+      case 8:
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldic, node, targetReg, sourceReg, 3,
+                                         CONSTANT64(0x00000007FFFFFFF8));
+         break;
+      case 4:
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldic, node, targetReg, sourceReg, 2,
+                                         CONSTANT64(0x00000003FFFFFFFC));
+         break;
+      case 2:
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, targetReg, sourceReg, sourceReg);
+         break;
+      case 1:
+         generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, node, targetReg, sourceReg);
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "immMultiplicatoinWithShift - unexpected referenceFieldSize!");
+      }
+   }
+
+static TR::Register *generateMultianewArrayWithInlineAllocators(TR::Node *node,
+                                                                 TR::CodeGenerator *cg,
+                                                                 int32_t leafArrayElementSize)
+   {
+   TR::Compilation *comp = cg->comp();
+   TR_Debug *compDebug = comp->getDebug();
+   TR_ASSERT_FATAL(comp->target().is64Bit(), "multianewArrayEvaluator is only supported on 64-bit JVMs!");
+   TR_J9VMBase *fej9 = comp->fej9();
+   // size of reference field and address
+   int32_t referenceFieldSize = TR::Compiler->om.sizeofReferenceField();
+   TR_ASSERT_FATAL(referenceFieldSize <= 8,
+      "multianewArrayEvaluator - referenceFieldSize cannot be greater than 8!");
+   int32_t addrSize = TR::Compiler->om.sizeofReferenceAddress();
+   // the size of a length-0 array in bytes taking alignment into account
+   int32_t zeroArraySizeAligned = OMR::align(TR::Compiler->om.discontiguousArrayHeaderSizeInBytes(),
+                                             TR::Compiler->om.getObjectAlignmentInBytes());
+   // we use 64 bit classes only in 64 bit arch without compression
+   bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
+   TR::InstOpCode::Mnemonic storeClassOp = use64BitClasses ? TR::InstOpCode::std : TR::InstOpCode::stw;
+   int8_t classPtrLen = use64BitClasses ? 8 : 4;
+
+   // Zero size arrays are considered "discontiguous", and the "mustBeZero" field
+   // of discontiguous arrays must be located where the "size" field of contiguous arrays is.
+   UDATA offsetOfMustBeZeroField = fej9->getOffsetOfContiguousArraySizeField();
+
+   // the maximum number of elements we can handle without risking an overflow
+   uintptr_t maxObjectSizeInElements = cg->getMaxObjectSizeGuaranteedNotToOverflow() / referenceFieldSize;
+
```

Comment:
as i commented before, the logic below cannot be followed easily.  why should we deal with objectAlignment requirement on referenceFieldSize? our ultimate goal is to calculate the (individual) final object (be it java array or normal object) sizes in such a way that each size is multiple of `TR::Compiler->om.getObjectAlignmentInBytes()`  (either 8 or 16 currently).

for each object,   `((headerSize + dataSize) + (alignmentInBytes - 1)) & (-alignmentInBytes)`, that is what in my mind, instead of looking at referenceFieldSize alone.  Basically, I don't understand how looking at fieldSize alone is related to fulfilling the final size alignment.

Reply:
that is obviously a wrong assumption and it can happen in major ways. example: gencon header is two slots (i.e. compressedRef mode, two slots are 8-byte). but, when heap size is bigger than 32GB, compression/decompression needs to shift by 4 (i.e. requiring 16-aligned). now, the header is not multiple of alignment anymore.

Reply:
looks like x/z existing implementations have multiple bugs:  leafElementSize mapping is one (except you only deal with 0-length array), and this header size assumption is another.